addons/mass_mailing/models/mailing_contact.py

```diff
@@ -38,6 +47,36 @@ def create(self, vals_list):
         return super().create(vals_list)
 
     def write(self, vals):
+        for record in self:
+            if 'opt_out' in vals or 'opt_out_reason' in vals:
+                vals['opt_out_reason'] = vals.get('opt_out', record.opt_out) and vals.get('opt_out_reason')
+            if 'opt_out' in vals or 'opt_out_custom' in vals:
+                vals['opt_out_custom'] = vals.get('opt_out', record.opt_out) and vals.get('opt_out_custom')
+            author_id = self.env.user.partner_id.id
+            if 'opt_out' in vals:
```

Comment:
I would somehow concatenate messages in a single one when having opt_our + reason. Otherwise it seems two messages are posted, which is probably not necessary.